packages/models-tree/src/models-tree/ModelsTreeIdsCache.ts

```diff
@@ -28,11 +29,18 @@ export class ModelsTreeIdsCache {
   private _subjectInfos: Promise<Map<Id64String, SubjectInfo>> | undefined;
   private _parentSubjectIds: Promise<Id64Array> | undefined; // the list should contain a subject id if its node should be shown as having children
   private _modelInfos: Promise<Map<Id64String, ModelInfo>> | undefined;
+  private _modelKeyPaths: Map<Id64String, HierarchyNodeIdentifiersPath[]>;
+  private _subjectKeyPaths: Map<Id64String, HierarchyNodeIdentifiersPath>;
+  private _categoryKeyPaths: Map<Id64String, HierarchyNodeIdentifiersPath[]>;
```

Comment:
These maps should store a Promise to value, otherwise we could end up with unnecessary calculations. E.g.:

```ts
async function getSubjectPath(id) {
    const curr = this._subjectKeyPaths.get(id);
    if (curr) {
        return curr;
    }
    const value = await doExpensiveCalculation(id);
    this._subjectKeyPaths.set(id, value);
    return value;
}

const [value1, value2] = await Promise.all(getSubjectPath(""), getSubjectPath("")));
// doExpensiveCalculation is now called twice!
```

Reply:
Changed, didn't think of that